sdk/core/core/src/az_client_secret_credential.c

```diff
@@ -43,125 +43,75 @@ static AZ_NODISCARD az_result az_token_credential_get_token(
 
-    auth_body = az_span_builder_result(&builder);
-  }
+    az_mut_span secret_copy = { .begin = builder.buffer.begin + builder.size, .size = 0 };
+    {
+      size_t secret_start = builder.size;
+      AZ_RETURN_IF_FAILED(az_uri_encode(credential->client_secret, &builder));
+      secret_copy.size = builder.size - secret_start;
+    }
 
-  {
-    assert(auth_url.size <= auth_url_maxsize);
-
-    az_http_request_builder hrb = { 0 };
-    AZ_RETURN_IF_FAILED(az_http_request_builder_init(
-        &hrb,
-        (az_mut_span){
-            .begin = response_buf.begin + response_buf.size - (auth_url.size),
-            .size = auth_url.size + headers_size,
+    az_result result = AZ_OK;
+    do {
+      AZ_BREAK_IF_FAILED(az_span_builder_append(&builder, auth_body3), result);
+      AZ_BREAK_IF_FAILED(az_uri_encode(resource_url, &builder), result);
+      auth_body = az_span_builder_result(&builder);
+
+      az_http_request_builder hrb = { 0 };
+      AZ_BREAK_IF_FAILED(
+          az_http_request_builder_init(
+              &hrb, response_buf, (uint16_t)auth_url.size, AZ_HTTP_METHOD_VERB_POST, auth_url),
+          result);
+
+      AZ_BREAK_IF_FAILED(az_http_request_builder_add_body(&hrb, auth_body), result);
+
+      static az_http_pipeline pipeline = {
+        .policies = {
+          { .pfnc_process = az_http_pipeline_policy_uniquerequestid, .data = NULL },
+          { .pfnc_process = az_http_pipeline_policy_retry, .data = NULL },
+          { .pfnc_process = no_op_policy, .data = NULL },
+          { .pfnc_process = az_http_pipeline_policy_logging, .data = NULL },
+          { .pfnc_process = az_http_pipeline_policy_bufferresponse, .data = NULL },
+          { .pfnc_process = az_http_pipeline_policy_distributedtracing, .data = NULL },
+          { .pfnc_process = az_http_pipeline_policy_transport, .data = NULL },
+          { .pfnc_process = NULL, .data = NULL },
         },
-        (uint16_t)auth_url.size,
-        AZ_HTTP_METHOD_VERB_POST,
-        auth_url));
-
-    AZ_RETURN_IF_FAILED(az_http_request_builder_add_body(&hrb, auth_body));
-
-    static az_http_pipeline pipeline = {
-    .policies = {
-      { .pfnc_process = az_http_pipeline_policy_uniquerequestid, .data = NULL },
-      { .pfnc_process = az_http_pipeline_policy_retry, .data = NULL },
-      { .pfnc_process = no_op_policy, .data = NULL },
-      { .pfnc_process = az_http_pipeline_policy_logging, .data = NULL },
-      { .pfnc_process = az_http_pipeline_policy_bufferresponse, .data = NULL },
-      { .pfnc_process = az_http_pipeline_policy_distributedtracing, .data = NULL },
-      { .pfnc_process = az_http_pipeline_policy_transport, .data = NULL },
-      { .pfnc_process = NULL, .data = NULL },
-    },
-    };
-
-    AZ_RETURN_IF_FAILED(az_http_pipeline_process(&hrb, &response_buf, &pipeline));
+      };
+
+      AZ_BREAK_IF_FAILED(az_http_pipeline_process(&hrb, &response_buf, &pipeline), result);
+    } while (false);
```

Comment:
What's the purpose a this do {} while(false).  
I think it is not a good approach because it is making you add an extra macro to BREAK on fail (extra complexity?)
Would it be the same to create a new method that uses `AZ_RETURN_IF_FAILED( )` and then we just call it from here?

That method can contain exactly what it is inside this do {} while(false)  and you can make it IN_LINE method to optimize it

Reply:
I rewrote the code in a way that I Ieft buffer allocation and cleanup in the parent function, and extracted the logic in between into a child function, where it is safe to use AZ_RETURN_IF_FAILED.